agents/agents-core/src/commonMain/kotlin/ai/koog/agents/core/agent/context/AIAgentContextBase.kt

```diff
@@ -180,4 +181,40 @@ public interface AIAgentContextBase {
         strategyId: String? = null,
         pipeline: AIAgentPipeline? = null,
     ): AIAgentContextBase
-}
\ No newline at end of file
+}
+
+
+/**
+ * Creates a deep copy of the AIAgentContextBase, ensuring that inner objects like llm and stateManager
+ * are properly copied to avoid shared state between contexts.
+ */
+@OptIn(ExperimentalUuidApi::class)
+internal suspend fun AIAgentContextBase.deepCopy(): AIAgentContextBase {
+    val newLlm = llm.readSession {
+        llm.copy(
+            tools = tools.toList(),
+            prompt = prompt.copy(),
+        )
+    }
+
+    val iterations = stateManager.withStateLock { it.iterations }
+    val newStateManager = AIAgentStateManager(AIAgentState(iterations))
+
+    return copy(
+        llm = newLlm,
+        stateManager = newStateManager,
+    )
+}
+
+/**
+ * Updates the current AI agent context by synchronizing its LLM prompt and state information
+ * with the provided new context.
+ */
+internal suspend fun AIAgentContextBase.update(newContext: AIAgentContextBase) {
+    val newPrompt = newContext.llm.readSession { prompt }
+    llm.writeSession { rewritePrompt { newPrompt } }
+
+    stateManager.withStateLock { state ->
+        state.iterations = newContext.stateManager.withStateLock { newState -> newState.iterations }
+    }
+}
```

Comment:
It looks to me like @tiginamaria is already adding something similar in https://github.com/JetBrains/koog/pull/220.